Implement swap without using scratch

utzig opened the pull request:
This implements a swap upgrade that does not use a scratch area. It works by first moving all sectors in the secondary slot up one position, and then looping on moving sector of index X of the primary slot to index X of the secondary slot, followed by moving sector X+1 of the secondary slot to X on the primary slot, for each sector X.

utzig commented:
Ah, also currently this allows for both `overwrite-only` and `swap-move` to be enabled at the same time, which makes overwrite-only the build option. Must also add a test + `#errror`

utzig commented:
Ok, now it's only missing design doc update. I will wait for review issues before I write it down.

utzig commented:
Btw, encrypted image swap is currently broken as was found out after merging e84f0effb!

d3zd3z commented on boot/bootutil/src/loader.c:
```diff
@@ -1111,16 +714,26 @@ boot_copy_region(struct boot_loader_state *state,
+            if (fap_dst->fa_id == FLASH_AREA_IMAGE_SECONDARY(image_index)) {
+                hdr = boot_img_hdr(state, BOOT_PRIMARY_SLOT);
+            }
+#endif
```
The problem I think exists in the code outside of this change, though. We do make an assumption that the header size is always less than our image size. The header size is generally defined by the size of the vector table on a particular device, which could be larger than 1k, which would result in corruption for just ordinary images (`blk_sz` would be negative).

utzig replied:
I never really thought of this possibility; I would suggest opening an issue!

utzig commented on boot/bootutil/src/loader.c:
```diff
@@ -1111,16 +714,26 @@ boot_copy_region(struct boot_loader_state *state,
+#else
+            off = off_dst;
+            if (fap_dst->fa_id == FLASH_AREA_IMAGE_SECONDARY(image_index)) {
+                hdr = boot_img_hdr(state, BOOT_PRIMARY_SLOT);
```
I added the header image checking (when we have enough information to do so), but the reason the test was failing was that there was no detection of inconsistent status which I have implemented now. With these latest changes I found that `perm_with_random_fails` can fail if exercised enough; not sure this was happening before. I would assume that swapping without scratch should be considered less robust than swap with scratch, at least until it is used and tested for some time.

d3zd3z commented on boot/bootutil/src/swap_move.c:
```diff
@@ -0,0 +1,506 @@
+        /* With validation of the primary slot disabled, there is no way
+         * to be sure the swapped primary slot is OK, so abort!
+         */
+        assert(0);
```
What happens on subsequent reset?  It seems like it would still go ahead and try running the new image again.

utzig replied:
This happens before any swap operation is started or resumed, so once a status area inconsistency is detected, it will be detected again on every new boot. I never really seen this happen in HW, and seems highly unlikely; we just try to handle as best as possible and do the testing in the simulator; but maybe add a `panic!` like macro that prints a message and loops forever?

nvlsianpu replied:
panic sound good to me

d3zd3z commented on boot/bootutil/src/swap_move.c:
```diff
@@ -0,0 +1,506 @@
+        /* This means there was an error writing status on the last
+         * swap. Tell user and move on to validation!
+         */
+        BOOT_LOG_ERR("Detected inconsistent status!");
```
I'm not sure what else we can do here, but if the status doesn't get written, what happens on the next boot?

utzig replied:
Same comment applies here...

nvlsianpu commented on boot/bootutil/src/bootutil_priv.h:
```diff
@@ -62,16 +62,24 @@ struct flash_area;
+#error "Please enable only one of MCUBOOT_OVERWRITE_ONLY or MCUBOOT_SWAP_MOVE"
+#endif
+
+#if !defined(MCUBOOT_OVERWRITE_ONLY) && !defined(MCUBOOT_SWAP_MOVE)
```
I prefer to state explicitly how to :
do swap using scratch algorithm  or -> MCUBOOT_SWAP_USING_SCRATCH
do swap using image sector shift algorithm -> MCUBOOT_SWAP_USING_MOVE
do not swap ->MCUBOOT_OVERWRITE_ONLY

utzig replied:
That is a good idea actually, I was unable to make my mind about the best naming to use!

utzig replied:
Done!

nvlsianpu commented on boot/bootutil/src/bootutil_priv.h:
```diff
@@ -62,16 +62,24 @@ struct flash_area;
 #define MCUBOOT_USE_SCRATCH 1
 #endif
 
+#define BOOT_STATUS_OP_MOVE     1
```
do you expect support more than one swap algorithm in any mcuboot build?

utzig replied:
No, why?

nvlsianpu replied:
I misunderstood purpose of these defines, as I reviewed their usage later. nevermind.

nvlsianpu commented on boot/bootutil/src/bootutil_priv.h:
```diff
@@ -179,7 +187,13 @@ _Static_assert(BOOT_IMAGE_NUMBER > 0, "Invalid value for BOOT_IMAGE_NUMBER");
 #endif
 
 /** Maximum number of image sectors supported by the bootloader. */
```
looks like this comment has peeled off what it describes

utzig replied:
Ah, this was already wrong before this PR, I will fix it.

utzig replied:
Should be fixed now!

nvlsianpu commented on boot/bootutil/src/swap_move.c:
```diff
@@ -0,0 +1,506 @@
+#define BOOT_STATUS_ASSERT(x) ASSERT(x)
+#endif
+
+static uint32_t g_last_idx = UINT32_MAX;
```
`g_` is this according to jull coding style?

utzig replied:
I don't remember we having a rule for globals, do you have a suggestion?

nvlsianpu replied:
... while this is module private. I don't have any suggestion, just guessing - but (i feel) I saw such convention in a recent jull patch to the mcumg.

nvlsianpu commented:
cc @hakonfam

nvlsianpu commented on boot/bootutil/src/swap_move.c:
```diff
@@ -0,0 +1,506 @@
+
+    rc = boot_write_status(state, bs);
+
+    bs->idx++;
```
I would increment `bs->idx` after return from call to `this`.
just a matter of taste, but I thing it will make algorithm more legible from higher level.

utzig replied:
I disagree; this function shares responsibilities with `boot_swap_sectors`, and they both modify the state internally. Would be weird if one modified the state while the other not.

nvlsianpu replied:
ok

nvlsianpu commented on boot/bootutil/src/swap_move.c:
```diff
@@ -0,0 +1,506 @@
+
+    /* FIXME: assure last sector does not overrun trailer */
+
+    if (bs->op == BOOT_STATUS_OP_MOVE) {
```
honestly: I would exclude flash_areas opening/closing and closing from `boot_move_sector_up()` and 'boot_swap_sectors()' and put it the opening before this `if` and the closing after the 'while' with `boot_swap_sectors` inside.
This would simplifies code of these sub-function a little, reduces footprint, speed up execution a little...

utzig replied:
Done with 41ebdfd69f3a1cda5e0fcfdc02693d615eba8d73. I will squash this commit into the original swap without scratch once approved.

nvlsianpu replied:
^^ patch looks good to me

oyvindronningstad commented on boot/bootutil/src/swap_move.c:
```diff
@@ -0,0 +1,480 @@
+ * a permanent upgrade in case a reset occurs.
+ */
+void
+fixup_revert(const struct boot_loader_state *state, struct boot_status *bs,
```
Can you add more explanation for this? E.g. what is the "right" time?

utzig replied:
It might be a bad wording; maybe it should be called "wrong" instead of "right". The idea is that there is a short period of time during the beginning of a revert where metadata might be lost because status in slot0 must be erased and status in slot1 was already previously erased. If a reset occurs at this time the swap breaks. The `fixup...` routine just saves temporarily the data in slot1 (and it will be erased again moments later), but it is saved long enough to walk a few extra steps without loosing information.

oyvindronningstad replied:
Thanks, I got what you meant by "right" ;). Your comment gives good context which I think should be added to the function doc comment.

utzig replied:
I did not copy the previous answer, but hopefully I improved the text with: 29ec1ab

oyvindronningstad commented on boot/bootutil/src/bootutil_priv.h:
```diff
@@ -238,14 +259,34 @@ int boot_read_swap_state(const struct flash_area *fap,
                          uint8_t image_num);
 int boot_write_swap_size(const struct flash_area *fap, uint32_t swap_size);
 int boot_read_swap_size(int image_index, uint32_t *swap_size);
+int boot_erase_trailer_sectors(const struct boot_loader_state *state,
```
Personally, I'd like the new swap implementations to have their own (shared) header file. To make it even more decoupled.

utzig replied:
Done with 5887060ac0f7636793b1a173693c9facb0e6e302. I did not go too far, simply moved new stuff from the PR, but it looks like it would be possible to move more non `overwrite-only` stuff in there... would rather do later...

utzig commented:
@oyvindronningstad I did not copy the previous answer, but hopefully I improved the text with: 29ec1aba203c050231edeea8a5072799511750de

utzig commented:
@oyvindronningstad I doubt I would be the best person to document this, all function names look self-explanatory to me! :joy:

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+
+#if defined(MCUBOOT_SWAP_USING_SCRATCH) || defined(MCUBOOT_SWAP_USING_MOVE)
+
+int boot_erase_trailer_sectors(const struct boot_loader_state *state,
```
Why does this need a separate function?

oyvindronningstad replied:
Why does this need a separate function?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+
+int boot_erase_trailer_sectors(const struct boot_loader_state *state,
+                               const struct flash_area *fap);
+int boot_copy_region(struct boot_loader_state *state,
```
Is this for copying between different slots or within the same slot?
Does it erase?
Does it have alignment restrictions?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+                     const struct flash_area *fap_src,
+                     const struct flash_area *fap_dst,
+                     uint32_t off_src, uint32_t off_dst, uint32_t sz);
+int boot_erase_region(const struct flash_area *fap, uint32_t off, uint32_t sz);
```
Does it have alignment restrictions?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+                     const struct flash_area *fap_dst,
+                     uint32_t off_src, uint32_t off_dst, uint32_t sz);
+int boot_erase_region(const struct flash_area *fap, uint32_t off, uint32_t sz);
+int boot_status_init(const struct boot_loader_state *state,
```
Does this initialize the structure in flash, or does it read from flash and initialize a RAM copy, or neither?
Which are input and output parameters?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+int boot_status_init(const struct boot_loader_state *state,
+                     const struct flash_area *fap,
+                     const struct boot_status *bs);
+void boot_perform_swap(struct boot_loader_state *state,
```
What are input and output parameters?
When is this meant to be called? Many times or once only?
Does it resume where it left off if the device resets?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+void boot_perform_swap(struct boot_loader_state *state,
+                       struct boot_status *bs,
+                       uint32_t copy_size);
+int boot_status_source(struct boot_loader_state *state);
```
What is a 'source' in this context? What are input and output parameters?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+                       struct boot_status *bs,
+                       uint32_t copy_size);
+int boot_status_source(struct boot_loader_state *state);
+int boot_read_status_bytes(const struct flash_area *fap,
```
What are 'status bytes' and where are they read from? Which parameter (or return value?) contains the status bytes after I've called this?

oyvindronningstad commented on boot/bootutil/src/bootutil_swap_priv.h:
```diff
@@ -0,0 +1,48 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Top level docs: e.g. Describe the end result of a swap, and maybe a sentence about each method (SCRATCH and MOVE)

utzig commented:
@oyvindronningstad I applied a few changes loosely based on your suggestions: renamed functions that are only required by swap upgrade functionality to `swap_` (from `boot_`); moved them to a new `swap_misc.c` file; added some extra documentation (very basic)  to `swap_priv.h` which does not substitute the required read of the design document. I will write more about how the swap without scratch works in the design document once this PR seems close to done (based on reviews).

utzig commented:
@oyvindronningstad Oops, forgot to point the commit, here it is: ba05ddc3520b25f96516316796b5b09078f25193

Laczen commented:
@utzig, nice that you are copying my idea for upgrade without scratch area from ZEPboot. The copy is not so good, you should first move up the image in the primary partition, this distributes the flash erases evenly. And it would be nice to mention the idea contributor.

utzig commented:
@Laczen Thanks, I've swapped the indexes and now it moves up the primary. `7a7b56fc125eb0d807ccb716ce4fb870c2c66838` was updated to mention the idea.

utzig commented:
@nvlsianpu @d3zd3z I added "[EXPERIMENTAL]" to the Zephyr Kconfig (https://github.com/JuulLabs-OSS/mcuboot/pull/589/commits/9d94fd3a8ca404951ec1ca9ef0132197e620b4c0#diff-56e9ade09681f44d80863bfce76a0f2bR130) and would like to leave it there at least until this looks as reliable as the scratch method. Is this OK?
